be/src/exprs/binary_predicate.cpp

```diff
@@ -151,49 +153,108 @@ class ArrayPredicate final : public Predicate {
     EvalCmpZero _comparator;
 };
 
-// add this class because VectorizedNullSafeEqPredicate can't deal with complex data type in columnViewer
-class VectorizedNullSafeArrayEqPredicate final : public Predicate {
+class CommonEqualsPredicate final : public Predicate {
 public:
-    explicit VectorizedNullSafeArrayEqPredicate(const TExprNode& node) : Predicate(node) {}
-    ~VectorizedNullSafeArrayEqPredicate() override = default;
+    explicit CommonEqualsPredicate(const TExprNode& node) : Predicate(node) {}
+    ~CommonEqualsPredicate() override = default;
 
-    Expr* clone(ObjectPool* pool) const override { return pool->add(new VectorizedNullSafeArrayEqPredicate(*this)); }
+    Expr* clone(ObjectPool* pool) const override { return pool->add(new CommonEqualsPredicate(*this)); }
+
+    StatusOr<ColumnPtr> evaluate_checked(ExprContext* context, Chunk* chunk) override {
+        ASSIGN_OR_RETURN(auto l, _children[0]->evaluate_checked(context, chunk));
+        ASSIGN_OR_RETURN(auto r, _children[1]->evaluate_checked(context, chunk));
+
+        if (l->only_null() || r->only_null()) {
+            return ColumnHelper::create_const_null_column(l->size());
+        }
+        auto& const1 = FunctionHelper::get_data_column_of_nullable(l);
+        auto& const2 = FunctionHelper::get_data_column_of_nullable(r);
+
+        size_t lstep = const1->is_constant() ? 0 : 1;
+        size_t rstep = const2->is_constant() ? 0 : 1;
+
+        auto& data1 = FunctionHelper::get_data_column_of_const(const1);
+        auto& data2 = FunctionHelper::get_data_column_of_const(const2);
```

Comment:
what about const(nullable col)? datax may be also nullable, and not_null_col->equals(nullable_column) is not well supported.

Reply:
const(nullable) is only_null

Reply:
and was handle `not_null_col->equals(nullable_column)`

Reply:
taking fixedLengthColumn as example, a no-nullable column does not process nullable columns.
```
template <typename T>
int FixedLengthColumnBase<T>::compare_at(size_t left, size_t right, const Column& rhs, int nan_direction_hint) const {
    DCHECK_LT(left, _data.size());
    DCHECK_LT(right, rhs.size());
    DCHECK(dynamic_cast<const FixedLengthColumnBase<T>*>(&rhs) != nullptr);
    T x = _data[left];
    T y = down_cast<const FixedLengthColumnBase<T>&>(rhs)._data[right];
    return SorterComparator<T>::compare(x, y);
}
```

Reply:
`const(nullable col) must be only_null` where guarantee this?

Reply:
I know, look line 182, I checked null on `l/r` first, then check data column on `data1->equals(loff, *(data2.get()), roff)`

Reply:
> 

it's a constraint when designed, if `const(nullable)` not `only_null`, why put a nullable column in const column not a data column?

Reply:
the key question is whether the data column of const columns would be nullable.

Reply:
BTW，maybe some places has break the constraint, we need add a `DCHECK` to promise the constraint...